program/freedom-metal-program.wake

```diff
@@ -1,127 +1,79 @@
-  h, t = match (splitTo f t)
-    Pair f s = Pair (h, f) s
-
-def addChosenSubnode = addSubnode `(L[a-zA-Z0-9_]+:\s)?chosen`
-
-def addSubnode parentNode node inDTSContents =
-  def split =
-    def firstSplit = splitTo (`^\t`, parentNode, `\s{$`, Nil).regExpCat.matches inDTSContents
-    def secondSplit = splitUntil `^\t};$`.matches firstSplit.getPairSecond
-    Pair (firstSplit.getPairFirst ++ secondSplit.getPairFirst) (secondSplit.getPairSecond)
-  split.getPairFirst ++ node ++ split.getPairSecond
-
-def addStdoutChosenFromAlias nodeName inDTSContents =
-  def parentNode = `(L[a-zA-Z0-9_]+:\s)soc`
-  def firstSplit = splitTo (`^\t`, parentNode, `\s{$`, Nil).regExpCat.matches inDTSContents
-  def secondSplit = splitUntil `\t\};$`.matches firstSplit.getPairSecond
-  def socContents = secondSplit.getPairFirst
-  def baudrate = 115200
-  match "({regExpToString nodeName.quote}@[0-9]+)".stringToRegExp
-    Fail e = inDTSContents
-    Pass nodeRegex =
-      def regex = regExpCat (`^\t\t(?:L[0-9]+:\s)?`, nodeRegex, `\s{$`, Nil)
-      def serialNodes = map regex.extract socContents | flatten
-      match serialNodes
-        serialNode, _ =
-          def stdChosen =
-            "\t\tstdout-path = \"/soc/{serialNode}:{str baudrate}\";", Nil
-          addChosenSubnode stdChosen inDTSContents
-        _ = inDTSContents
-
-def addChosenNode inDTSContents =
-  def parentNode = "/"
-  def firstSplit = splitTo (parentNode.quote, `\s{$`, Nil).regExpCat.matches inDTSContents
-  def secondSplit = splitUntil `.*{.*`.matches firstSplit.getPairSecond
-  def chosenNode = "\tchosen \{", "\t\};", Nil
-
-  firstSplit.getPairFirst
-  ++ secondSplit.getPairFirst
-  ++ chosenNode
-  ++ secondSplit.getPairSecond
-
-global target fixupDTS dts stdoutPath metalEntryPairOpt romNode ramNode =
-  def clkFrequency = 32500000
-  def prefix =
-    extract `(.*)\.dts` dts.getPathName
-    | head
-    | getOrElse dts.getPathName
-
-  match dts.read
-    Pass dtsContents =
-      dtsContents
-      | tokenize `\n`
-      | addChosenNode
-      | (omap addStdoutChosenFromAlias stdoutPath | getOrElse (_))
-      | (omap addEntry metalEntryPairOpt | getOrElse (_))
-      | (omap "metal,rom".addChosenRef romNode | getOrElse (_))
-      | (omap "metal,ram".addChosenRef ramNode | getOrElse (_))
-      | catWith "\n"
-      | write "{prefix}_fixed.dts"
-    fail = dts
+      match `RISCV_ABI`.getAttribute `RISCV_ARCH`.getAttribute `RISCV_CMODEL`.getAttribute
+        (Some mabi) (Some march) (Some cmodel) = Pass (FreedomMakeAttributes attributes march mabi cmodel)
+        _ _ _ = Fail "Generated make attributes file not formatted correctly: {attributes.getPathName}".makeError
+    Fail fail = Fail fail
+
+tuple FreedomBSP =
+  global LinkerScript:   Path
+  global Header:         Path
+  global InlineHeader:   Path
+  global PlatformHeader: Path
+  global Attributes:     FreedomMakeAttributes
+
+global def makeFreedomBSP coreDTS prefix =
+  def type = "rtl"
```

Comment:
The overlay generator (and attributes/esdk settings generator, but it's less critical here) needs a `type` argument from the list of
- rtl
- arty
- vc707
- qemu
- hifive
- spike

This tells the overlay generator which heuristics to follow when picking which memories to use and more. Do we want to expose this as a parameter to `makeFreedomBSP`, and if so, how should that me made available for consumers of `api-generator-sifive`?

Reply:
Pushed a commit to pass in FPGABoardOpt if it isn't None, otherwise default to "rtl".